### ISSUE TITLE
`collapsible_match` FP when scrutinee could match following arm

### ISSUE THREAD

**Issue by narpfel** [rust-lang/rust-clippy#16717]
### Summary

The transformation suggested by `collapsible_match` alters the behaviour when the scrutinee could match a later pattern.

### Lint Name

collapsible_match

### Reproducer

I tried this code:

```rust
#![allow(unused)]

enum E {
    A(&'static str, bool),
    B(&'static str),
    C,
}

impl E {
    fn name(&self) -> Option<&'static str> {
        match self {
            Self::A(name, _) | Self::B(name) => Some(name),
            Self::C => None,
        }
    }
}

fn f(e: E) {
    match e {
        E::A(_, update) => {
            if update {
                println!("side effect: update");
            }
        }
        _ if let Some(name) = e.name() => println!("other side effect: {name}"),
        _ => (),
    }
}

pub fn main() {
    f(E::A("no output expected", false));
}
```

I saw this happen:

```
warning: this `if` can be collapsed into the outer `match`
  --> <source>:21:13
   |
21 | /             if update {
22 | |                 println!("side effect: update");
23 | |             }
   | |_____________^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#collapsible_match
   = note: `#[warn(clippy::collapsible_match)]` on by default
help: collapse nested if block
   |
20 ~         E::A(_, update)
21 ~             if update => {
22 |                 println!("side effect: update");
23 ~             }
   |

warning: 1 warning emitted
```

I expected to see this happen: The lint should not be triggered.

In the original code, *any* `E::A(_, _)` will execute the first `match` arm (and possibly not execute the side effect (if `update` is `false`)).

With the autofix, `E::A(_, false)` will *not* match the first arm, so the next arm will be tried and successfully matches.

Godbolt link: https://godbolt.org/z/hTcscGj5P

### Version

```text
rustc 1.96.0-nightly (03749d625 2026-03-14)
binary: rustc
commit-hash: 03749d625404994801804ab9f2a539936c50bfea
commit-date: 2026-03-14
host: x86_64-unknown-linux-gnu
release: 1.96.0-nightly
LLVM version: 22.1.0
```

**Comment by ronnodas** [comment]
Is it better to not trigger the lint or to suggest the following?
```rust
E::A(_, update) if update => {...}
E::A(_, update) => {}
```

I would actually prefer here:
```rust
E::A(_, true) => {...}
E::A(_, false) => {}
```
But that should probably be a separate issue.

**Comment by profetia** [comment]
Fixed in #16878
